### PR TITLE
Automatic batching cost functions

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: theseus/core/objective.py [hunk 1]
```diff
@@ -107,6 +107,27 @@ def _get_cost_function_variable_batch_name(var: Manifold):
             + f"{tuple(var.data.shape[1:])}"
         )
 
+    @staticmethod
+    def _get_cost_function_batch_name(cost_function: CostFunction) -> Union[str, None]:
```
Comment (fantaosha) [inline theseus/core/objective.py:111]: @luisenp @mhmukadam You don't have to review this PR.  If you have time, please check if this function generates unique batch name for cost functions under the assumption that all the `optim_vars` and `aux_vars` have the same batch size.

File: theseus/core/objective.py [hunk 2]
```diff
@@ -109,6 +107,29 @@ def _get_cost_function_variable_batch_name(var: Manifold):
...
+            return None
+
+        batch_name = cost_function.__module__ + "." + cost_function.__class__.__name__
+
+        variable_batch_names = [
+            "-" + Objective._get_cost_function_variable_batch_name(var)
+            for var in variables
+        ]
+
+        return batch_name + "".join(variable_batch_names)
```
Comment (mhmukadam) [inline theseus/core/objective.py:131]: Does this lead to long string like you were worried about before?

File: theseus/core/objective.py [hunk 3]
```diff
@@ -109,6 +107,29 @@ def _get_cost_function_variable_batch_name(var: Manifold):
...
+        variables.extend([aux_var for aux_var in cost_function.aux_vars])
+        batch_sizes = [variable.shape[0] for variable in variables]
+        unique_batch_sizes = set(batch_sizes)
+
+        if len(unique_batch_sizes) != 1 and (
+            len(unique_batch_sizes) != 2 or min(unique_batch_sizes) != 1
+        ):
+            return None
+
+        batch_name = cost_function.__module__ + "." + cost_function.__class__.__name__
```
Comment (mhmukadam) [inline theseus/core/objective.py:124]: Is module name needed?

File: theseus/core/objective.py [hunk 4]
```diff
@@ -109,6 +107,29 @@ def _get_cost_function_variable_batch_name(var: Manifold):
...
+        variables: List[Variable] = [
+            optim_var for optim_var in cost_function.optim_vars
+        ]
+        variables.extend([aux_var for aux_var in cost_function.aux_vars])
+        batch_sizes = [variable.shape[0] for variable in variables]
+        unique_batch_sizes = set(batch_sizes)
+
+        if len(unique_batch_sizes) != 1 and (
+            len(unique_batch_sizes) != 2 or min(unique_batch_sizes) != 1
+        ):
```
Comment (mhmukadam) [inline theseus/core/objective.py:121]: Didn't follow this logic. Maybe add a comment about what's happening here?

File: theseus/core/objective.py (hunk 1 above)
Comment (mhmukadam) [reply theseus/core/objective.py:111]: LGTM!

**Comment by fantaosha** [comment]
Merged to `experimental`.

**Pull request closed without merging**